synapse/util/ratelimitutils.py

```diff
@@ -133,12 +133,16 @@ def ratelimit(self, host: str) -> "Iterator[defer.Deferred[None]]":
         self.host = host
 
         request_id = object()
-        ret = self._on_enter(request_id)
+        ret = defer.ensureDeferred(self._on_enter_with_tracing(request_id))
         try:
             yield ret
         finally:
             self._on_exit(request_id)
 
+    async def _on_enter_with_tracing(self, request_id: object) -> None:
+        with start_active_span("ratelimit wait"), queue_wait_timer.time():
+            await self._on_enter(request_id)
```

Comment:
the only thing about this is that we'll now start the tracing spans even if we hit the ratelimit and raise a `LimitExceededError` (at line 156).

That's probably fine, but a way around it would be to call `self._on_enter` (and stash the deferred in a local var) *before* the `with` stuff.

Reply:
I gave that a go, but it turns out that we end up in a logging context-less void after `_on_enter` has returned, where it's not legal to start a span.

Reply:
oh, right. ugh. thanks.